larpandoracontent/LArTwoDReco/TrackInEMShowerAlgorithm.cc

```diff
@@ -516,224 +510,188 @@ const Cluster* TrackInEMShowerAlgorithm::RefineTrack(const Cluster *const pClust
     }
 
     // End fragmentation
-    PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::EndFragmentation(*this, fragmentListName, originalListName)); 
-    
-    for (const Cluster *const pCreatedCluster : allCreatedClusters)
+    if (pAboveCluster || pBelowCluster)
+    {
+        PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::EndFragmentation(*this, fragmentListName, originalListName));
+    }
+    else
     {
-        if (pCreatedCluster->GetNCaloHits() == 1)
-            this->AddToNearestCluster(pCreatedCluster, pClusterToEnlarge);
+        PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::EndFragmentation(*this, originalListName, fragmentListName));
     }
-
-    // ATTN: Do not add clusters that now pass the 'clean cluster' threshold into the cluster vector, since these are unlikely to be a muon track
+            
 }
 
 //------------------------------------------------------------------------------------------------------------------------------------------
 
-void TrackInEMShowerAlgorithm::MergeHits(const ClusterAssociation &clusterAssociation, const ClusterToCaloHitListMap &clusterToCaloHitListMap,
-    TwoDSlidingFitResultMap &microSlidingFitResultMap, TwoDSlidingFitResultMap &macroSlidingFitResultMap,  ClusterVector &clusterVector) const
+bool TrackInEMShowerAlgorithm::IsClusterRemnantDisconnected(const Cluster *const pRemnantCluster) const
 {
-    // Merge together cosmic ray muon pair
-    const Cluster *const pClusterToEnlarge(clusterAssociation.GetUpstreamCluster());
-    const Cluster *const pClusterToDelete(clusterAssociation.GetDownstreamCluster());
-
-    this->UpdateForClusterDeletion(pClusterToDelete, clusterVector, microSlidingFitResultMap, macroSlidingFitResultMap);    
-    PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::MergeAndDeleteClusters(*this, pClusterToEnlarge, pClusterToDelete));
+    if (pRemnantCluster->GetNCaloHits() == 1)
+        return false;
 
-    // Merge in all collected hits from shower region between the pair
-    ClusterVector clustersToFragment;
-    for (const auto &entry : clusterToCaloHitListMap)
-        clustersToFragment.push_back(entry.first);
+    unsigned int count(0);
+    float maxSeparationDistance(-std::numeric_limits<float>::max());
 
-    // In the case of a delta ray split
-    if (clustersToFragment.empty())
+    const OrderedCaloHitList &orderedCaloHitList(pRemnantCluster->GetOrderedCaloHitList());
+    CartesianVector previousHitPosition((*orderedCaloHitList.begin()->second->begin())->GetPositionVector());
+    for (const OrderedCaloHitList::value_type &mapEntry : orderedCaloHitList)
     {
-        this->UpdateAfterClusterModification(pClusterToEnlarge, clusterVector, microSlidingFitResultMap, macroSlidingFitResultMap);
-        return;
-    }
+        for (const CaloHit *const pCaloHit : *mapEntry.second) 
+        {
+            ++count;
+            
+            if (count == 1)
```

Comment:
This confused me at first, but it's just to avoid comparing hit with itself the first time round. Is it clearer to keep track of `pPreviousCaloHit`, rather than `previousHitPosition`, then can always just skip over cases where `pPreviousCaloHit == pCaloHit`?

Reply:
much clearer, thank you!!